### DISCUSSION TITLE
how to save and load best model &how to auto select gpu/cpu

### DISCUSSION THREAD

**Discussion by lg2578** [Lightning-AI/pytorch-lightning#13076] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
I've used keras, pytorch ,and I found pytorch-lightning is amazing.
my questions are:
### 1.how to save and load best model ?
I've tried code  like this:
```
from pytorch_lightning.callbacks import ModelSummary,ModelCheckpoint
cb_savebest=ModelCheckpoint(monitor="val_acc",mode='max',every_n_epochs=1,filename='best')
trainer=pl.Trainer(
       accelerator="gpu",gpus=1,
        max_epochs=10,
        callbacks=[ModelSummary(max_depth=1),],
        limit_train_batches=10,
                  )
model=Lit_Classifier()
trainer.fit(model,train_dl,val_dl)
bestmodel=cb_savebest.best_model_path
model.load_from_checkpoint(bestmodel)

```
which get an error  :[Errno 21] Is a directory: '/kaggle/working', 
why best_model_path not a filename?
2.how to auto select gpu/cpu?
`trainer=pl.Trainer(
       accelerator="gpu",gpus=1,`
the above code works when there is a gpu,
but if I turn off the gpu, I need to comment the line 
` accelerator="gpu",gpus=1,`
is there a way like
device = torch.device(" cuda " if torch.cuda. is_available () else " cpu ")

**Comment by lg2578** [comment]
question 2 is solved:
```
trainer = Trainer(
    accelerator="auto",
    devices=1 if torch.cuda.is_available() else None,  # limiting got iPython runs
    max_epochs=3,
    callbacks=[TQDMProgressBar(refresh_rate=20)],
)
```

**Comment by rohitgr7** [comment]
can you share the stack trace for the first error?
